integration-libs/opf/b2b-checkout/root/guards/opf-b2b-checkout-steps-set.guard.ts

```diff
@@ -0,0 +1,186 @@
+          const stepRouteUrl = `/${
+            this.routingConfigService.getRouteConfig(step.routeName)?.paths?.[0]
+          }`;
+          return stepRouteUrl === currentRouteUrl;
+        });
+        // get current step
+        let currentStep;
+        if (currentIndex >= 0) {
+          currentStep = steps[currentIndex];
+        }
+        if (Boolean(currentStep)) {
+          return this.isB2BStepSet(steps[currentIndex - 1], isAccount);
+        } else {
+          return of(this.getUrl('checkout'));
+        }
+      })
+    );
+  }
+
+  protected isOpfCheckoutFlow(): Observable<boolean> {
+    return this.checkoutFlowOrchestratorService
+      .getPaymentProvider()
+      .pipe(map((provider) => provider === OPF_CHECKOUT_FLOW_NAME));
+  }
+
+  protected isB2BStepSet(
+    step: CheckoutStep,
+    isAccountPayment: boolean
+  ): Observable<GuardResult> {
+    if (step && !step.disabled) {
+      switch (step.type[0]) {
+        case CheckoutStepType.PAYMENT_TYPE: {
+          return this.isPaymentTypeSet(step);
+        }
+        case CheckoutStepType.DELIVERY_ADDRESS: {
+          return this.isDeliveryAddressAndCostCenterSet(step, isAccountPayment);
+        }
+        case CheckoutStepType.DELIVERY_MODE: {
+          return this.isDeliveryModeSet(step);
+        }
+        case CheckoutStepType.PAYMENT_DETAILS: {
+          return this.isPaymentDetailsSet(step);
+        }
+        case CheckoutStepType.REVIEW_ORDER: {
+          break;
+        }
+      }
+    }
+    return of(true);
+  }
+
+  protected isPaymentTypeSet(step: CheckoutStep): Observable<GuardResult> {
+    return this.checkoutPaymentTypeFacade.getSelectedPaymentTypeState().pipe(
+      filter((state) => !state.loading),
+      map((state) => state.data),
+      map((paymentType) => {
+        if (paymentType) {
+          return true;
+        } else {
+          return this.getUrl(step.routeName);
+        }
+      })
+    );
+  }
+
+  protected isDeliveryAddressAndCostCenterSet(
+    step: CheckoutStep,
+    isAccountPayment: boolean
+  ): Observable<GuardResult> {
+    return combineLatest([
+      this.checkoutDeliveryAddressFacade.getDeliveryAddressState().pipe(
+        filter((state) => !state.loading),
+        map((state) => state.data)
+      ),
+      this.checkoutCostCenterFacade.getCostCenterState().pipe(
+        filter((state) => !state.loading),
+        map((state) => state.data)
+      ),
+    ]).pipe(
+      map(([deliveryAddress, costCenter]) => {
+        if (isAccountPayment) {
+          if (
+            deliveryAddress &&
+            Object.keys(deliveryAddress).length &&
+            !!costCenter
+          ) {
+            return true;
+          } else {
+            return this.getUrl(step.routeName);
+          }
+        } else {
+          if (deliveryAddress && Object.keys(deliveryAddress).length) {
+            return true;
+          } else {
+            return this.getUrl(step.routeName);
+          }
+        }
```

Comment:
Could this be simplified?  Something more like this?
```
        const addressKeys = Object.keys(deliveryAddress ?? {}).length;
        if (isAccountPayment) {
          if (addressKeys && !!costCenter) {
            return true;
          }
        } else {
          if (addressKeys) {
            return true;
          }
        }
        return this.getUrl(step.routeName);
```

Reply:
It is the same case as in: https://github.com/SAP/spartacus/pull/20210#discussion_r2079196788